### PR TITLE
Add `MinimizeRegisterPressure` strategy to `LoopInvariantOptimization`

### PR DESCRIPTION
**Pull request by larsk21**
+ correctly call `confirmProperties(NONE)` in `LinearBlocksOptimization`

### CODE REVIEW

File: src/main/java/edu/kit/compiler/optimizations/LoopInvariantOptimization.java [hunk 1]
```diff
@@ -2,43 +2,62 @@
...
+        Map<Block, Set<Block>> loops = loopInvariantAnalysis.getLoops();
         Map<Block, List<Node>> loopInvariantNodes = loopInvariantAnalysis.getLoopInvariantNodes();
 
-        MoveInvariantStrategy moveInvariantStrategy = new MoveInvariantStrategies.MoveAlways();
+        boolean backEdgesEnabled = BackEdges.enabled(graph);
+        if (!backEdgesEnabled) {
+            BackEdges.enable(graph);
+        }
+
+        MoveInvariantStrategy moveInvariantStrategy = this.moveInvariantStrategy.orElse(new MoveInvariantStrategies.MinimizeRegisterPressure(loops, loopInvariantNodes));
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/LoopInvariantOptimization.java:57]: This line is a bit long^^

File: src/main/java/edu/kit/compiler/optimizations/NodeCosts.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            case iro_Mul:
+                return 4;
+            case iro_Mulh:
+                return 4;
+            case iro_Div:
+                return 25;
+            case iro_Mod:
+                return 25;
+            case iro_Alloc:
+                return 50;
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/NodeCosts.java:48]: Alloc usually involves multiple loads, and would thus probably be even more expensive

File: src/main/java/edu/kit/compiler/optimizations/NodeCosts.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            case iro_Load:
+                return 50;
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/NodeCosts.java:50]: Assuming we might have an L1-cache hit in many cases, 50 is a bit high of an estimate, but should be ok for our needs

File: src/main/java/edu/kit/compiler/optimizations/NodeCosts.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            case iro_Store:
+                return 50;
+            case iro_Call:
+                return 2;
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/NodeCosts.java:54]: Shouldn't the cost estimate for a call include the execution of the whole function behind the call (and thus be very high)?

File: src/main/java/edu/kit/compiler/optimizations/loop_invariant/MoveInvariantStrategies.java [hunk 5]
```diff
@@ -19,10 +39,176 @@ private MoveInvariantStrategies() {}
...
+         * a block to the immediate dominator of the block. If the immediate
+         * dominator of a block is not one of its direct predecessors, then the
+         * step is counted twice.
+         */
+        private int moveDistance(Node node, Block target) {
+            int distance = 0;
+
+            Block block = (Block) node.getBlock();
+            // try to move to the target block, but never further
+            while (!block.equals(target) && Util.findDeepestCommonDominator(Arrays.asList(block, target)).equals(target)) {
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/loop_invariant/MoveInvariantStrategies.java:180]: this line is also a bit long

**Review by N-Maas — COMMENTED** [review]
Generally, the heuristic looks good to me

File: src/main/java/edu/kit/compiler/optimizations/NodeCosts.java (hunk 3 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/NodeCosts.java:50]: True, I looked it up again. Interesting how fast caches have gotten.

File: src/main/java/edu/kit/compiler/optimizations/NodeCosts.java (hunk 4 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/NodeCosts.java:54]: Ideally yes, but there is no reasonable estimate for the execution time of a function. Of course, the heuristic treats function calls as expensive.

**Review by N-Maas — APPROVED** [review]

**Pull request merged**